README.md
# Description

  Apia is a Haskell program for proving first-order theorems written
  in [Agda](http://wiki.portal.chalmers.se/agda/pmwiki.php) using
  automatic theorem provers for first-order logic (ATPs). Before
  calling the ATPs, the Agda formulae are translated into
  [TPTP](http://www.cs.miami.edu/~tptp/) language.

# Use of Apia

  * Reasoning about Functional Programs by Combining Interactive and
    Automatic Proofs
    ([README.md](https://github.com/asr/fotc/blob/master/README.md)).

# Prerequisites

* Extended version of Agda

  We have extended the development version of Agda in order to handle
  the new built-in ATP-pragma. This extended version of Agda is
  required by the Apia program.

* ATPs

  Apia requires at least one of the following ATPs:
  [E](http://www4.informatik.tu-muenchen.de/~schulz/WORK/eprover.html),
  [Equinox](http://www.cse.chalmers.se/~koen/code/),
  [ileanCoP](http://www.leancop.de/ileancop/index.html),
  [Metis](http://www.gilith.com/software/metis/),
  [SPASS](http://www.spass-prover.org/) or
  [Vampire](http://www.vprover.org/). The tested versions of the ATPs
  are: E 1.8-001 Gopaldhara, Equinox 5.0alpha (2010-06-29), ileanCoP
  1.3beta1, Metis 2.3 (release 20141206), SPASS 3.7 and Vampire 0.6
  (revision 903).

* The tptp4X program

  The `--check` command-line option requires the tptp4X program from
  the [TPTP library](http://www.cs.miami.edu/~tptp/). The tested
  version of tptp4X is from TPTP 6.1.0.

# Installation

1. Extended version of Agda (see
   [README.md](https://github.com/asr/eagda/blob/master/README.md))

2. The Apia program

   You can download the Apia program using
   [Git](http://git-scm.com/). The program can be downloaded and
   installed with the following commands:

   ````bash
   $ git clone https://github.com/asr/apia.git
   $ cd apia
   $ cabal install
   ````

   In order to test the installation of the program, once the extended
   version of Agda and (some of) the ATPs have been installed, we can
   try to automatically prove the conjecture in

    ````agda
    module Test where

    data _∨_ (A B : Set) : Set where
      inj₁ : A → A ∨ B
      inj₂ : B → A ∨ B

    postulate
      A B    : Set
      ∨-comm : A ∨ B → B ∨ A
    {-# ATP prove ∨-comm #-}
    ````

   by running the following commands:

   ````bash
   $ agda Test.agda
   $ apia Test.agda
   Proving the conjecture in /tmp/Test/9-8744-comm.tptp ...
   E 1.8-001 Gopaldhara proved the conjecture
   ````

   The program will call the default ATPs and tell which of the ATPs
   was able to first prove a certain conjecture. If none ATP could
   prove a conjecture after four minutes, the process of proving that
   particular conjecture is aborted and the ATPs try to prove the next
   conjecture.

   If we want to just use a certain ATP, say Equinox, we can instead
   run the following command:

   ````bash
   $ apia --atp=equinox Test.agda
   Proving the conjecture in /tmp/Test/9-8744-comm.tptp ...
   Equinox, version 5.0alpha, 2010-06-29 proved the conjecture
   ````

# Known limitations

* Logical symbols

  The following symbols are hard-coded, i.e. they should be used: `⊥`
  (falsehood), `⊤` (truth), `¬_` (negation), `_∧_` (conjunction),
  `_∨_` (disjunction), the Agda non-dependent function type `→`
  (implication), `_↔_` (equivalence), the Agda dependent function type
  `(x : A) → B` (universal quantifier), `∃` (existential quantifier),
  and `_≡_` (propositional equality).

* Agda version

  The Apia program must be compiled using the same version of Agda
  that was used to generate the Agda interface files.


old/CHANGES.txt
NB. This file should be add in the .cabal file:
extra-source-files:
        CHANGES.txt


Change log for Apia

0.9.1 XXX

      * Added command-line option --snapshot-no-error.

      * The translation of FOL universal quantified functions,
        formulae, and propositional functions is disable by
        default. We added to Agda the pragmas

        --universal-quantified-functions,
        --universal-quantified-formulae and
        --universal-quantified-propositional-functions

        for this purpose. The Apia program is handling these
        pragmas.

      * Added command-line option --unproven-conjecture-no-error.

      * Removed command-line option --unproven-conjecture-error. Now
        it is the default behaviour.

0.9 April 2012
    Start of changelog


src/Apia/ATPs.hs
-----------------------------------------------------------------------------
-- |
-- Module      : Apia.ATPs
-- Copyright   : (c) Andrés Sicard-Ramírez 2009-2014
-- License     : See the file LICENSE.
--
-- Maintainer  : Andrés Sicard-Ramírez <asr@eafit.edu.co>
-- Stability   : experimental
--
-- Call the automatic theorem provers for first-order logic (ATPs).
-----------------------------------------------------------------------------

{-# LANGUAGE CPP                 #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE UnicodeSyntax       #-}

module Apia.ATPs
  ( ATP  -- Required by Haddock.
  , callATPs
  ) where

------------------------------------------------------------------------------
-- Haskell imports

import Control.Applicative     ( (<$>) )
import Control.Exception       ( evaluate )
import Control.Concurrent      ( forkIO )
import Control.Concurrent.MVar ( MVar, newEmptyMVar, putMVar, takeMVar )
import Control.Monad           ( when )
import Control.Monad.IO.Class  ( MonadIO(liftIO) )

import Data.List  ( isInfixOf )
import Data.Maybe ( fromMaybe, isNothing )

import Safe ( initDef )

import System.Directory ( findExecutable )
import System.IO        ( hGetContents )

import System.Process
  ( CmdSpec(RawCommand)
  , createProcess
  , CreateProcess
    ( close_fds
    , create_group
    , CreateProcess
    , cmdspec
    , cwd
#if MIN_VERSION_process(1,2,0)
    , delegate_ctlc
#endif
    , env
    , std_err
    , std_in
    , std_out
    )
  , interruptProcessGroupOf
  , ProcessHandle
  , readProcess
  , StdStream(CreatePipe, Inherit)
  )

------------------------------------------------------------------------------
-- Agda library imports

import Agda.Utils.Impossible ( Impossible(Impossible) , throwImpossible )
import Agda.Utils.Monad      ( ifM )

------------------------------------------------------------------------------
-- Apia imports

import Apia.Monad.Base    ( askTOpt, T )
import Apia.Monad.Reports ( reportS )

import Apia.Options ( Options(optATP, optTime, optUnprovenNoError, optVampireExec) )

import qualified Apia.Utils.Except as E

#include "undefined.h"

------------------------------------------------------------------------------
-- | The ATPs.
data ATP = E
         | Equinox
         | IleanCoP
         | Metis
         | SPASS
         | Vampire
           deriving Show

atpExec ∷ ATP → T String
atpExec E        = return "eprover"
atpExec Equinox  = return "equinox"
atpExec IleanCoP = return "ileancop.sh"
atpExec Metis    = return "metis"
atpExec SPASS    = return "SPASS"
atpExec Vampire  = askTOpt optVampireExec

optATP2ATP ∷ String → T ATP
optATP2ATP "e"        = return E
optATP2ATP "equinox"  = return Equinox
optATP2ATP "ileancop" = return IleanCoP
optATP2ATP "metis"    = return Metis
optATP2ATP "spass"    = return SPASS
optATP2ATP "vampire"  = return Vampire
optATP2ATP other      = E.throwE $ "ATP " ++ other ++ " unknown"

-- | Default ATPs.
defaultATPs ∷ [String]
defaultATPs = ["e", "equinox", "vampire"]

atpOk ∷ ATP → String
-- E 1.2, E 1.3, E 1.4, E 1.5, E 1.6, E 1.7 and E 1.8.
atpOk E = "Proof found!"
-- Equinox 5.0alpha (2010-06-29).
atpOk Equinox = "+++ RESULT: Theorem"
-- ileanCoP 1.3 beta1.
atpOk IleanCoP = "Intuitionistic Theorem"
-- Metis 2.3 (release 20141206).
atpOk Metis = "SZS status Theorem"
-- SPASS 3.7.
atpOk SPASS = "Proof found"
-- Vampire 0.6 (revision 903).
atpOk Vampire = "Termination reason: Refutation\n"

atpVersion ∷ ATP → T String
 -- No version option in Equinox.
atpVersion Equinox = do
  exec ← atpExec Equinox
  liftIO $ fmap (initDef (__IMPOSSIBLE__) . takeWhile (/= '\n'))
                (readProcess exec ["--help"] "")
-- No version option in ileanCoP.
atpVersion IleanCoP = return $ show IleanCoP
-- No version option in SPASS.
atpVersion SPASS = return $ show SPASS
atpVersion atp = do
  exec ← atpExec atp
  liftIO $ initDef (__IMPOSSIBLE__) <$> readProcess exec ["--version"] ""

checkOutput ∷ ATP → String → Bool
checkOutput atp output = atpOk atp `isInfixOf` output

atpArgs ∷ ATP → Int → FilePath → T [String]
atpArgs E timeLimit file = do
  eVersion ← atpVersion E
  if eVersion `elem` [ "E 1.2 Badamtam"
                     , "E 1.3 Ringtong"
                     , "E 1.4 Namring"
                     , "E 1.5 Pussimbing"
                     ]
    then return [ "--cpu-limit=" ++ show timeLimit
                , "--expert-heuristic=Auto"
                , "--memory-limit=Auto"
                , "--output-level=0"
                , "--term-ordering=Auto"
                , "--tstp-format"
                , file
                ]
    else
      if eVersion `elem` [ "E 1.6 Tiger Hill"
                         , "E 1.7 Jun Chiabari"
                         , "E 1.8-001 Gopaldhara"
                         ]
        then return [ "--auto"
                    , "--cpu-limit=" ++ show timeLimit
                    , "--memory-limit=Auto"
                    , "--output-level=0"
                    , "--tstp-format"
                    , file
                    ]
        -- This message is not included in the error test.
        else E.throwE $ "the ATP " ++ eVersion ++ " is not supported"

-- Equinox bug. Neither the option @--no-progress@ nor the option
-- @--verbose 0@ reduce the output.
atpArgs Equinox timeLimit file = return [ "--time", show timeLimit
                                        , file
                                        ]

-- N.B. The order of the ileanCoP arguments is fixed.
atpArgs IleanCoP timeLimit file = return [ file
                                         , show timeLimit
                                         ]

atpArgs Metis timeLimit file = return [ "--time-limit", show timeLimit
                                      , file
                                      ]

atpArgs SPASS timeLimit file = return [ "-PProblem=0"
                                      , "-PStatistic=0"
                                      , "-TimeLimit=" ++ show timeLimit
                                      , "-TPTP=1"
                                      , file
                                      ]

-- 25 July 2012. We don't know if Vampire has an option to reduce the
-- output.
atpArgs Vampire timeLimit file = return ["--mode", "casc"
                                        , "-t", show timeLimit
                                        , "--input_file", file
                                        ]

runATP ∷ ATP → MVar (Bool, ATP) → Int → FilePath → T ProcessHandle
runATP atp outputMVar timeLimit file = do
  args ∷ [String] ← atpArgs atp timeLimit file
  cmd ∷ String    ← atpExec atp

  e ← liftIO $ findExecutable cmd
  when (isNothing e) $ E.throwE $
    "the command " ++ cmd ++ " associated with " ++ show atp
    ++ " does not exist.\nYou can use the command-line option --atp=NAME "
    ++ "to avoid call some ATP"

  -- To create the ATPs process we follow the ideas used by
  -- @System.Process.proc@.
  (_, outputH, _, atpPH) ← liftIO $
    createProcess CreateProcess
                    { cmdspec       = RawCommand cmd args
                    , cwd           = Nothing
                    , env           = Nothing
                    , std_in        = Inherit
                    , std_out       = CreatePipe
                    , std_err       = Inherit
                    , close_fds     = False
                    , create_group  = True
#if MIN_VERSION_process(1,2,0)
                    , delegate_ctlc = False
#endif
                    }
  output ← liftIO $ hGetContents $ fromMaybe (__IMPOSSIBLE__) outputH
  _      ← liftIO $ forkIO $
             evaluate (length output) >>
             putMVar outputMVar (checkOutput atp output, atp)

  return atpPH

atpsAnswer ∷ [String] → MVar (Bool, ATP) → [ProcessHandle] → FilePath → Int →
             T ()
atpsAnswer atps outputMVar atpsPH file n =
  if n == length atps
    then do
      let msg ∷ String
          msg = "the ATP(s) did not prove the conjecture in " ++ file
      ifM (askTOpt optUnprovenNoError)
          (liftIO $ putStrLn msg)
          (E.throwE msg)
    else do
      output ← liftIO $ takeMVar outputMVar
      atpWithVersion ← atpVersion (snd output)
      if fst output
        then do
          reportS "" 1 $ atpWithVersion ++ " proved the conjecture"
          liftIO $ mapM_ interruptProcessGroupOf atpsPH
        else do
          reportS "" 1 $ atpWithVersion ++ " *did not* prove the conjecture"
          atpsAnswer atps outputMVar atpsPH file (n + 1)

-- | The function 'callATPs' calls the selected 'ATP's on a TPTP conjecture.
callATPs ∷ FilePath → T ()
callATPs file = do
  atpsAux       ← askTOpt optATP
  timeLimitAux  ← askTOpt optTime
  outputMVar    ← liftIO (newEmptyMVar ∷ IO (MVar (Bool, ATP)))

  let atps ∷ [String]
      atps = if null atpsAux then defaultATPs else atpsAux

  reportS "" 1 $ "Proving the conjecture in " ++ file
  reportS "" 20 $ "ATPs to be used: " ++ show atps

  -- See note [Timeout increse].
  let timeLimit ∷ Int
      timeLimit = round (fromIntegral timeLimitAux * (1.1 ∷ Float))

  atpsPH ∷ [ProcessHandle] ←
    mapM optATP2ATP atps >>= mapM (\atp → runATP atp outputMVar timeLimit file)

  atpsAnswer atps outputMVar atpsPH file 0

------------------------------------------------------------------------------
-- Note [Timeout increse].

-- 12 June 2012: Hack. Running for example
--
-- @$ equinox --time 216 conjecture.tptp@
--
-- or
--
-- @$ apia --time=216 --atp=equinox conjecture.agda@
--
-- it is possible prove the theorem. But running for example
--
-- @$ apia --time=216 --atp=equinox --atp=vampire --atp=e conjecture.agda@
--
-- doesn't prove the theorem. I guess there is some overhead for
-- calling various ATPs from Apia. Therefore we increase internally
-- 10% the ATPs timeout.
